### ISSUE TITLE
Static typing support for typeguard

### ISSUE THREAD

**Issue by bfelbo** [microsoft/pyright#614]
**Is your feature request related to a problem? Please describe.**
We have values obtained from an untrusted source, which we need to check conforms to our types at runtime. We only do simple checking like string literals. For this we use the popular [typeguard](https://github.com/agronholm/typeguard). However, pyright doesn't detect that after running typeguard's `check_type` on a value, then the value must necessarily be that type. Consequently, we get type errors even though those type errors would never occur at runtime.

**Describe the solution you'd like**
I'd like the following code snippet to not report any pyright issues.
```
from typeguard import check_type
from typing_extensions import Literal

Gender = Literal["Male", "Female"]
def print_gender(gender: Gender):
    print(gender)

# Unknown string as it's been retrieved from elsewhere
strRetrievedFromDB = "Male"  # type: str

# As check_type() ensures the str is complying with the type,
# pyright should not report any issues, but it currently does.
check_type("Gender", strRetrievedFromDB, Gender)
print_gender(strRetrievedFromDB)
```

**Comment by bfelbo** [comment]
Alternatively, can you give any advice on how to write a simple type check function that will work with pyright?

Obviously, the code below won't work, but something as simple as this does work with pyright would be great.
```
def check_string_literal(value: str, expected_type: Literal):
    assert value in expected_type
```

**Comment by erictraut** [comment]
What you're looking for is something that's referred to as ["user-defined type guards"](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards). Unfortunately, the Python type spec has no provision for this currently.

The best alternative that I can offer you is to write a wrapper around typeguard's check_type that returns the value once it validates the type. Something like this:
```
_T = TypeVar("_T")

def validate_type(argname: str, value: Any, expected_type: _T) -> _T:
    # check_type will raise an exception if the type doesn't match
    check_type(argname, value, Gender)
    return cast(expected_type, value)

# Unknown string as it's been retrieved from elsewhere
strRetrievedFromDB = "Male" # type: str

validatedStr = validate_type("Gender", strRetrievedFromDB, Gender)
print_gender(validatedStr)
```

**Comment by bfelbo** [comment]
That works. Thanks for the quick response!
